### ISSUE TITLE
Support calling influxql/flux 'selector' aggregates from IOx SQL

### ISSUE THREAD

**Issue by alamb** [influxdata/influxdb_iox#5528]
# background

InfluxQL and Flux have specialized timeseries functions called selectors: `first` and `last`
https://docs.influxdata.com/flux/v0.x/stdlib/universe/first/
https://docs.influxdata.com/flux/v0.x/stdlib/universe/last/
https://docs.influxdata.com/influxdb/v1.8/query_language/functions/#first
https://docs.influxdata.com/influxdb/v1.8/query_language/functions/#last

These selectors behave like aggregates in the sense that they take many rows in and produce a single row out

However, they are somewhat specialized as they rely on the value of the timestamp and return the value associated with the earliest and latest time, respectively

In order to support influxrpc ReadAggregate and ReadWindowAggregate IOx implements datafusion extension aggregate functions for this functionality. However, they can not be called from SQL directly (only programatically by the influxrpc --> LogicalPlan conversion logic). 

The flux team (@Marwes  and @wolffcm specifically) are working on implementing Flux in terms of IOx SQL, and to properly push down these selectors, it would be very helpful to call them via SQL

# Details

Basically to get first(host) the query made by the influxrpc frontend looks like

```sql
SELECT selector_first_value(host, time) as host,
       selector_first_time(time) as time
  FROM table;
```

And then there is a custom implementation of `selector_first_value` that tracks the values corresponding to the first timestamp.

IOx implementation:

https://github.com/influxdata/influxdb_iox/blob/b820f69b0552ad4225044d84300edcae4301056d/query_functions/src/selectors.rs#L37-L51

Internal slack context: https://influxdata.slack.com/archives/CRK9M8L5Q/p1661957034784189

**Comment by jacobmarble** [comment]
The cited code mentions a JIRA feature request, which was migrated to GH/arrow-df as https://github.com/apache/arrow-datafusion/issues/600 . Does that Arrow feature request block this IOx feature request?

**Comment by alamb** [comment]
@jacobmarble 

The issue is that a "selector" must return two values: the field's value as well as the corresponding timestamp. SQL aggregates return a single value. 

# Option 1
One possible way of modeling this is to return a single "struct" value that has two fields, aka what https://github.com/apache/arrow-datafusion/issues/600 describes

This would allow you to write SQL like this (accessing `.value` and `.time` fields)
```sql
SELECT 
  f.value, 
  f.time 
FROM 
  (select first(col, time) as f FROM ..)
```

Or maybe
```sql
SELECT 
  first(col, time).value
  first(col, time).time
FROM ...
```

# Option 2

The workaround I came up with was to create two separate aggregate functions that compute part of the selector. This is the approach the influxrpc frontend uses to handle `ReadGroup` and `ReadWindowAggregate` with selectors. 

This requires you to write SQL like this (not super dissimilar to the above)

```sql
SELECT 
  first_value(col, time) as col,
  first_time(col, time) as time,
FROM table;
```

**Comment by wolffcm** [comment]
I've got a few comments here.

> The issue is that a "selector" must return two values: the field's value as well as the corresponding timestamp. SQL aggregates return a single value.

This is true of influxrpc but it gets a little more complicated when dealing with data that is "pivoted" (i.e. plain old normal tables). Here's an example.

### Selectors with just one field

Consider this schema where `host` and `region` are tags, and `cpu` and `memory`  are fields:
| host | region | cpu | memory | time |
|----|----|----|----|----|

If we run a Flux statement like this:
```js
from(...) |> range(...)
  |> filter(fn: (r) => r._field == "cpu")
  |> group(columns: ["region"])
  |> max()
```
Then we expect to get one row for each region, where the columns in that row should include not only the timestamp but also the `host` from the row that had the max `cpu`. In other words, it's not just `time` that needs to "come along for the ride".

So in order to push down both the `group` and the selector, I think we'd need a solution that is a bit more general. This will be a common case for us.

If we can do the selector without the `group` to start with, only needing to bring `time` along, that would be a good start. Also, we know the table schema on the Flux side (unlike when we are talking to influx rpc) so we can specify each column explicitly on the SELECT list as needed.

### Things get weird when there are more fields being selected

As an aside, If we remove the `_field` predicate, we end up with a different situation, since influxrpc will execute the selector on each series. That is to say, each field must have its max value selected for each unique combination of `(host, region)`.
```js
from(...) |> range(...)
  |> group(columns: ["region"])
  |> max()
```
What we'd really want from this  Flux query might look something like this coming out of IOx:

| host | region | max_cpu | max_memory | time |
|----|----|----|----|----|
| host01 | us | 75.0 | *null* | 2022-09-02T00:00:23Z |
| host01 | us | *null* | 12349.0 | 2022-09-01T00:03:01Z |
| host02 | eu | 76.0 | *null* | 2022-09-01T00:12:23Z |
| host02 | eu | *null* | 123475.0 | 2022-09-02T09:00:23Z |

In other words, we need to be able to compute the `max` value for each field independently.

I realize that this it _very_ different from how things work in SQL, and I'm not expecting any special support for this from IOx. (Maybe we need to craft a SQL statement that uses UNION with a separate SELECT for each field.) I'm just including this as an example of what we're dealing with in making IOx table data look like influxrpc data.

**Comment by alamb** [comment]
Hi @wolffcm -- this is a great writeup. Thank you for that -- for your example of getting `max` from multiple fields, the way the influxrpc frontend in IOx would do it today that we would get separate time columns for each distinct field.

So for example, that would result in something like (sorry my markdown skills are messing up the formatting):

| host | region | max_cpu | max_memory | time_max_cpu | time_max_memory |
|----|----|----|----|----|----|----|
| host01 | us | 75.0 | 12349.0 | 2022-09-02T00:00:23Z | 2022-09-01T00:03:01Z |
| host02 | eu | 76.0 | 123475.0 | 2022-09-01T00:12:23Z | 2022-09-02T09:00:23Z |

Would that work for Flux?

If not, then your idea to use `UNION ALL` is likely the way to go

**Comment by wolffcm** [comment]
@alamb My sense is that using a selector with multiple fields happens mostly because it's a Flux query that's easy to write by accident, and not because it's actually what users want to do. So I think it would be fair to consider it low priority until there's a reason otherwise.

The most general case (several fields, with a call to `group` in the flux query) would have to produce something more like this for `group(columns: ["region"])`:
| host_max_cpu | host_max_memory | region | max_cpu | max_memory | time_max_cpu | time_max_memory |
|----|----|----|----|----|----|----|

I guess the number of "along for the ride" columns would be on the order of `num_fields * num_nongroupkey_cols` which doesn't really seem practical? `UNION ALL` is looking better to me now.

**Comment by alamb** [comment]
I think as long as the number of columns less than ~100 it will be fine. Above that I suspect we will see some scale challenges.

But in any event I'll get the selector's exposed in SQL and then we can figure out the best way to invoke them from Flux

**Comment by stuartcarnie** [comment]
I was curious what the value of the `time` column will be, if you execute two selectors (`min` and `max`) in a single InfluxQL query.

### Data contains a single row

**Insert data[^1]**

```
insert cpu val=100
```

[^1]: Using `influx` v1.x client

**Query**

```sql
SELECT min(val), max(val) FROM cpu
```

**Output**

Note that the `time` column is `0`, despite both values having the same timestamp[^2]. 

[^2]: This is the same behaviour if the rows have differing timestamps.

```text
name: cpu
time max min
---- --- ---
0    100 100
```

**Comment by alamb** [comment]
> I was curious what the value of the time column will be, if you execute two selectors (min and max) in a single InfluxQL query.

@stuartcarnie the TLDR is that  the main observation is that there is no "single" `time` column -- you'll get a time coumn out for each of the selectors you want to run. 

So instead of 
```sql
SELECT min(val), max(val) FROM cpu
```

The query would look like

```sql
SELECT min_val(val, time), min_time(val, time),  max_val(val, time), max_time(val, time) FROM cpu;
```

Resulting in 


| min_val  | min_time | max_val  | max_time |
| ------------- | ------------- | ------------- | ------------- |
| 100  | 0  | 100  | 0  |



> Similarly, min, max, last, top and bottom should all be able to use WINDOW functions.

Yes, I think that is correct. However, I am fairly sure the performance of the WINDOW function implementation will be quite a bit worse than the selector version. Among other reasons, the current implementation of WINDOW in datafusion will sort the entire intermediate result set (though it does attempt to group the PARTITION BY clauses to minimize the number of searches).

> I believe @alamb has talked about adding a SQL rewrite feature to DataFusion, which would be awesome, as I expect we could use that to add virtual functions for these selectors, which would simplify the DataFusion SQL statements. Using the rewriter, we could transform them to use WINDOW functions, where appropriate.

Yes, the approach I propose is basically https://github.com/influxdata/influxdb_iox/pull/5438 which is a rewrite on the parsed SQL query.

**Comment by stuartcarnie** [comment]
> The query would look like
> 
> ```sql
> SELECT min_val(val, time), min_time(val, time),  max_val(val, time), max_time(val, time) FROM cpu;
> ```
> 

I like the rewritten query, as it does more flexibility 👍🏻 

Does Flux return the independent time values for the max and min? InfluxQL does not, because as soon as you add more than one selector, the timestamp returned is always `0` given there is only a single `time` column returned. With that in mind, for InfluxQL, we can rewrite it to:

```sql
SELECT 0 as time, min_val(val, time),  max_val(val, time) FROM cpu;
```

> 
> Yes, the approach I propose is basically #5438 which is a rewrite on the parsed SQL query.

This is great 👏🏻

**Comment by alamb** [comment]
> Does Flux return the independent time values for the max and min?

Yes -- and it can do so because it returns "timeseries" rather than a table, and each series has its own timestamps (aka logical set of timestamps)

**Comment by alamb** [comment]
I have a PR ready to rock https://github.com/influxdata/influxdb_iox/pull/5628
